File: panda-monitor-backend/src/main/kotlin/cn/zzwtsy/panda/controller/ServerGroupController.kt
package cn.zzwtsy.panda.controller

import cn.zzwtsy.panda.Result
import cn.zzwtsy.panda.model.ServerGroup
import cn.zzwtsy.panda.model.dto.ServerGroupInput
import cn.zzwtsy.panda.model.name
import org.babyfish.jimmer.sql.kt.KSqlClient
import org.babyfish.jimmer.sql.kt.ast.expression.eq
import org.slf4j.LoggerFactory
import org.springframework.transaction.annotation.Transactional
import org.springframework.web.bind.annotation.*

/**
 * @author zzwtsy
 */
@RestController
@RequestMapping("/serverGroup")
class ServerGroupController(private val kSqlClient: KSqlClient) {
    private val log = LoggerFactory.getLogger(javaClass)

    /**
     * 添加服务器分组
     *
     * @param input 服务器分组信息
     */
    @PostMapping
    @Transactional(rollbackFor = [Throwable::class])
    fun addServerGroup(input: ServerGroupInput): Result<ServerGroup> {
        val isExist = getServerGroup(input.name)
        if (isExist != null) {
            log.warn("分组【${input.name}】已存在，无法添加")
            return Result.error(400, "分组【${input.name}】已存在，无法添加")
        }
        val result = kSqlClient.save(input).modifiedEntity
        return Result.success(result)
    }

    /**
     * 删除服务器分组
     *
     * @param serverName 服务器分组名称
     */
    @DeleteMapping("/{serverName}")
    @Transactional(rollbackFor = [Throwable::class])
    fun deleteServerGroup(@PathVariable serverName: String): Result<String> {
        val isExist = getServerGroup(serverName)
        if (isExist == null) {
            log.warn("分组【${serverName}】不存在，无法删除")
            return Result.error(400, "分组【${serverName}】不存在，无法删除")
        }
        return Result.success()
    }

    /**
     * 修改服务器分组名称
     *
     * @param input 服务器分组信息
     */
    @PutMapping
    @Transactional(rollbackFor = [Throwable::class])
    fun updateServerGroup(input: ServerGroupInput): Result<ServerGroup> {
        val isExist = getServerGroup(input.name)
        if (isExist == null) {
            log.warn("分组【${input.name}】不存在，无法修改")
            return Result.error(400, "分组【${input.name}】不存在，无法修改")
        }
        val result = kSqlClient.update(input).modifiedEntity
        return Result.success(result)
    }

    /**
     * 获取服务器分组列表
     */
    @GetMapping("/list")
    fun listServerGroup(): Result<List<ServerGroup>> {
        val result = kSqlClient.createQuery(ServerGroup::class) {
            select(table)
        }.execute()
        return Result.success(result)
    }

    private fun getServerGroup(name: String): ServerGroup? {
        return kSqlClient.createQuery(ServerGroup::class) {
            where(table.name eq name)
            select(table)
        }.fetchOneOrNull()
    }
}

File: panda-monitor-web/src/__generated/services/ServerGroupController.ts
import type { Executor } from '../'
import type { Dynamic_ServerGroup } from '../model/dynamic/'
import type { Result, ServerGroupInput } from '../model/static/'

export class ServerGroupController {
  constructor(private executor: Executor) {}

  /**
   * 添加服务器分组
   *
   * @param input 服务器分组信息
   */
  readonly addServerGroup: (
    options: ServerGroupControllerOptions['addServerGroup']
  ) => Promise<Result<Dynamic_ServerGroup>> = async (options) => {
    let _uri = '/serverGroup'
    let _separator = _uri.indexOf('?') === -1 ? '?' : '&'
    let _value: any = undefined
    _value = options.input.name
    _uri += _separator
    _uri += 'name='
    _uri += encodeURIComponent(_value)
    _separator = '&'
    return (await this.executor({ uri: _uri, method: 'POST' })) as Promise<
      Result<Dynamic_ServerGroup>
    >
  }

  /**
   * 删除服务器分组
   *
   * @param serverName 服务器分组名称
   */
  readonly deleteServerGroup: (
    options: ServerGroupControllerOptions['deleteServerGroup']
  ) => Promise<Result<string>> = async (options) => {
    let _uri = '/serverGroup/'
    _uri += encodeURIComponent(options.serverName)
    return (await this.executor({ uri: _uri, method: 'DELETE' })) as Promise<Result<string>>
  }

  /**
   * 获取服务器分组列表
   */
  readonly listServerGroup: () => Promise<Result<ReadonlyArray<Dynamic_ServerGroup>>> =
    async () => {
      let _uri = '/serverGroup/list'
      return (await this.executor({ uri: _uri, method: 'GET' })) as Promise<
        Result<ReadonlyArray<Dynamic_ServerGroup>>
      >
    }

  /**
   * 修改服务器分组名称
   *
   * @param input 服务器分组信息
   */
  readonly updateServerGroup: (
    options: ServerGroupControllerOptions['updateServerGroup']
  ) => Promise<Result<Dynamic_ServerGroup>> = async (options) => {
    let _uri = '/serverGroup'
    let _separator = _uri.indexOf('?') === -1 ? '?' : '&'
    let _value: any = undefined
    _value = options.input.name
    _uri += _separator
    _uri += 'name='
    _uri += encodeURIComponent(_value)
    _separator = '&'
    return (await this.executor({ uri: _uri, method: 'PUT' })) as Promise<
      Result<Dynamic_ServerGroup>
    >
  }
}

export type ServerGroupControllerOptions = {
  addServerGroup: {
    /**
     * 服务器分组信息
     */
    readonly input: ServerGroupInput
  }
  deleteServerGroup: {
    /**
     * 服务器分组名称
     */
    readonly serverName: string
  }
  updateServerGroup: {
    /**
     * 服务器分组信息
     */
    readonly input: ServerGroupInput
  }
  listServerGroup: {}
}
